File: CSCI382/c++/MaxHeap.cpp
/* An object implementation of a max heap that supports sorting and priority queues. */

/* the priority queue stuff still needs testing */

#include <iostream>
#include <vector>
#include <string>
#include <cstdlib>
using namespace std;

struct HeapElement
{
    void *handle;   // points to an application object. remember void pointers must be explicitly cast.
    int key;
};

class MaxHeap
{
    private:
        int heapsize;
        vector<HeapElement *> heap;
        
        static int Parent(int i);
        static int Left(int i);
        static int Right(int i);

        void Swap(int i, int j)
        {
            HeapElement * c = heap[i];
            heap[i] = heap[j];
            heap[j] = c;
        }
        void MaxHeapify(int i);
        void BuildMaxHeap();
        
    public:
        MaxHeap(vector<HeapElement *> &);
        void HeapSort();
        
        // getter methods for heap and heapsize
        int getheapsize() const {return heapsize;}
        vector<HeapElement *> getheap() const {return heap;}

        // these methods support usage of MaxHeap objects as priority queues
        void* Max() {return heap[0]->handle;}
        void* ExtractMax();
        void IncreaseKey(int i, int newKey);
        void Insert(HeapElement* newElement);
        
        
};

int MaxHeap::Parent(int i) { return (i+1)/2-1; } // implicit floor division, adjust for index order 0
int MaxHeap::Left(int i) { return 2*(i+1)-1; } 
int MaxHeap::Right(int i) { return 2*(i+1)+1-1; }

void MaxHeap::MaxHeapify(int i)
{
    int l = Left(i);
    int r = Right(i);
    int largest;
    if ( l < heapsize && heap[l]->key > heap[i]->key ) largest = l;
    else largest = i;
    if ( r < heapsize && heap[r]->key > heap[largest]->key ) largest = r;
    if ( largest != i )
    {
        Swap(i, largest);
        MaxHeapify(largest);
    }
}

void MaxHeap::BuildMaxHeap()
{
    for ( int i = heapsize/2-1; i >= 0; i-- ) MaxHeapify(i);
}

MaxHeap::MaxHeap(vector<HeapElement*> &v)
{
    heapsize = v.size();
    heap = v;
    BuildMaxHeap();
}

void MaxHeap::HeapSort()
{
    while (heapsize > 0)
    {
        Swap(0, heapsize-1);
        heapsize--;
        MaxHeapify(0);
    }
    // let's reset heapsize to the length of heap (a vector)
    heapsize = heap.size();
}

void* MaxHeap::ExtractMax()
{
    if (heapsize < 1)
    {
        cout << "Heap underflow." << endl;
        exit(EXIT_FAILURE);
    } else {
        void* max = heap[0]->handle;
        delete heap[0]; //free the memory pointed to by the extracted pointer
        heap[0] = heap[heapsize-1]; //copy the last pointer to the first
        heap.pop_back(); //remove the old copy
        heapsize--;
        MaxHeapify(0);
        return max;
    }
}

void MaxHeap::IncreaseKey(int i, int newKey)
{
    if (heap[i]->key >= newKey)
    {
        cout << "New key must be greater than current key." << endl;
        exit(EXIT_FAILURE);
    } else {
        heap[i]->key = newKey;
        while (i > 0 && heap[i]->key > heap[Parent(i)]->key)
        {
            Swap(i, Parent(i));
            i = Parent(i);
        }
    }
}

void MaxHeap::Insert(HeapElement* newElement)
{
    // save the element's key so we can set it to -1 for IncreaseKey
    int key = newElement->key;
    // -1 is less than every possible key, so IncreaseKey won't break.
    newElement->key = -1;
    heapsize++;
    if (heapsize > heap.size()) {
        heap.push_back(newElement);
    } else {
        heap[heapsize-1] = newElement;
    }
    IncreaseKey(heapsize-1, key);
}

int main(int argc, char **argv)
{
    // these two lines are the best way I've come up with for putting command line arguments
    //  into a vector of HeapElement pointers.
    vector<HeapElement *> argHE;
    for (int i=0; i < argc-1; i++) argHE.push_back(new HeapElement);
    for (int i=0; i < argc-1; i++){
        argHE[i]->key = atoi(argv[i+1]);
        argHE[i]->handle = &(argHE[i]->key); // setting handle (a void*) equal to the key address (an int*)
    }                                         // just to test things out
    MaxHeap h = argHE; 
    for (int i=0; i<h.getheapsize(); i++) {
        cout << "key: " << h.getheap()[i]->key << "\t" << "handle: " << *((int*)h.getheap()[i]->handle) << endl;
    }
    cout << endl;
    HeapElement* e = new HeapElement;
    e->key = 3;
    e->handle = &e->key;
    h.Insert(e);
    for (int i=0; i<h.getheapsize(); i++) {
        cout << "key: " << h.getheap()[i]->key << "\t" << "handle: " << *((int*)h.getheap()[i]->handle) << endl;
    }
    
    // The delete below will fail if h.heap contains degenerate pointers (ptrs that point to the same
    // HeapElement object). This shouldn't happen in practice, since different HeapElements should have
    // different keys, even if they contatain ptrs to the same application object. However, the modern
    // workaround for this is to use shared_ptr<T> instead. Since you can't cast shared_ptr<void> to
    // shared_ptr<int> the same way you can cast void* to anything, I'd have to modify  my hackney example
    // to make HeapElement::handle a shared_ptr<int> specifically.
    // Ideally I'd make an ApplicationObject class and then make HeapElement::handle a 
    // shared_ptr<ApplicationObject>.
    for (int i=0; i< h.getheap().size(); i++) {
        delete h.getheap()[i];
    }
}

File: CSCI382/python/MaxHeap.py
"""
An object implementation of a max heap that support sorting and priority queues.
For now, each heap element is equal to its key. To implement a priority queue, we
need to create a Heap Element object that contains the key and the handle.
"""

# heap elements should contain a key and a handle
class HeapElement():
    def __init__(self, key, handle=None):
        self.key = key
        self.handle = handle

# takes a list of heap elements, and returns a max heap.
class MaxHeap():
    def __init__(self, heap):
        self.heapsize = len(heap)
        self.heap = heap
        self.BuildMaxHeap()

    # the following staticmethods are defined a bit differently than
    #   in CLRS since CLRS uses index order 1 and we python uses 0.
    @staticmethod
    def Parent(i):
        return (i+1)//2-1

    @staticmethod
    def Left(i):
        return 2*(i+1)-1

    @staticmethod
    def Right(i):
        return 2*(i+1)+1-1

    def MaxHeapify(self, i):
        l = MaxHeap.Left(i)
        r = MaxHeap.Right(i)
        # lists are mutable so we can nickname self.heap H for brevity
        H = self.heap
        # unfortunately we can't nickname heapsize since scalars are immutable
        if l < self.heapsize and H[l].key > H[i].key:
            largest = l
        else:
            largest = i
        if r < self.heapsize and H[r].key > H[largest].key:
            largest = r
        if largest != i:
           H[i], H[largest] = H[largest], H[i]
           self.MaxHeapify(largest)

    def BuildMaxHeap(self):
        i = self.heapsize//2 - 1
        while i >= 0:
            self.MaxHeapify(i)
            i-=1

    def HeapSort(self):
        # lists are mutable so we can nickname self.heap H for brevity
        H = self.heap
        while self.heapsize > 1:
            H[0].key, H[self.heapsize-1].key = H[self.heapsize-1].key, H[0].key
            self.heapsize -= 1
            self.MaxHeapify(0)
        # we have to set heapsize back to the length of the heap now
        self.heapsize = len(self.heap)

    # returns a handle corresponding to the application object with the largest key
    # runtime: O(1)
    def Max(self):
        return self.heap[0].handle

    # returns a handle to the element of the max heap with the largest key.
    # runtime: O(logn)
    def ExtractMax(self):
        if self.heapsize < 1:
            print "Heap underflow."
            return -1
        MAX = self.heap[0].handle
        self.heap[0] = self.heap[self.heapsize-1]
        self.heapsize -= 1
        self.MaxHeapify(0)
        return MAX

    # Increases the key of the heap element at index i to "key"
    # runtime: O(logn)
    def IncreaseKey(self, i, key):
        H = self.heap
        if H[i].key >= key:
            print "New key must exceed current key."
            return -1
        H[i].key = key
        while i > 0 and H[i].key > H[MaxHeap.Parent(i)].key:
            H[i], H[MaxHeap.Parent(i)] = H[MaxHeap.Parent(i)], H[i]
            i = MaxHeap.Parent(i)

    # Inserts a new HeapElement object at the appropriate key. Uses IncreaseKey().
    # Tacks it onto the end of the heap and then increases it to the right key.
    #runtime: O(logn)
    def Insert(self,element):
        # we assume key is always positive. then -1 is always less than key.
        # these two lines make sure IncreaseKey doesn't throw an error.
        key = element.key
        element.key = -1

        self.heapsize += 1
        # check to see if there's "empty space" in the heap. if not, append. if so just assign.
        # note this will potentially overwrite elements that were previously extracted by
        # ExtractMax().
        if self.heapsize > len(self.heap):
            self.heap.append(element)
        else:
            # overwrites a previously extracted element
            self.heap[self.heapsize-1] = element
        self.IncreaseKey(self.heapsize-1,key)


"""
# let's test it out:
A = [HeapElement(x,x) for x in range(0,10)]
H = MaxHeap(A)
H.Insert(HeapElement(10,"Hello World!"))
print H.heapsize
print len(H.heap)
print H.Max()
"""
